### PR TITLE
Initial version of Diameter Accountung RFC 7155

### PR DESCRIPTION
**Pull request by surik**
This PR adds an initial version of Diameter client to send Accounting data according to RFC 7155.

Example of diameter provider configuration:
```
{ergw_aaa_provider,
    {ergw_aaa_diameter,
        [{nas_identifier, <<"nas01.example.com">>},
         {host,    <<"example.com">>}},
         {realm,   <<"client.example.com">>}},
         {connect_to,     <<"aaa://10.10.10.10:3868">>}}
        ]}
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
-- 19.1
+    - 19.1
+    - 19.2
+    - 19.3
+before_install:
+    - wget https://github.com/rebar/rebar/wiki/rebar
+    - chmod u+x ./rebar
+    - wget https://s3.amazonaws.com/rebar3/rebar3
+    - chmod u+x ./rebar3
+env:
```
Comment (RoadRunnr) [inline .travis.yml]: I think rebar3 only CI test are sufficient, please sync this with gtplib's, see https://github.com/travelping/gtplib/commit/5ed158eda5b96c8c4913f1f9db7b4c6e7eb165b0 (.travis.yml, rebar.config.* and README.md) in one cset, drop the rebar2 support, and make the minimal version R19

And make this a separate PR to we can apply it independent of the diameter changes.

File: dia/diameter_nasreq_rfc7155.dia [hunk 2]
```diff
@@ -0,0 +1,341 @@
+@id	1
```
Comment (RoadRunnr) [inline dia/diameter_nasreq_rfc7155.dia]: Please adjust to formatting the to match that of the DIAMETER dictionaries shipped with OTP, see https://github.com/erlang/otp/blob/master/lib/diameter/src/dict/base_rfc6733.dia.

Add a copyright header that is consistent with the other files in this project,
 e.g. https://github.com/travelping/ergw_aaa/blob/master/src/ergw_aaa.erl#L1

File: dia/diameter_nasreq_rfc7155.dia [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+Accounting-Input-Packets                     	365      	Unsigned64          	M   
+Accounting-Output-Packets                    	366      	Unsigned64          	M   
+NAS-Filter-Rule                              	400      	IPFilterRule        	M   
+Tunneling                                    	401      	Grouped             	M   
+Accounting-Auth-Method                       	406      	Enumerated          	M   
+QoS-Filter-Rule                              	407      	QoSFilterRule       	M   
+Origin-AAA-Protocol                          	408      	Enumerated          	M   
+
+@messages
+AC-Request ::= <Diameter Header: 271, REQ, PXY>
```
Comment (RoadRunnr) [inline dia/diameter_nasreq_rfc7155.dia]: commonly ACR is used for this. Is there a reason not to use ACR (maybe a collision with /base_rfc6733.dia?)

File: dia/diameter_nasreq_rfc7155.dia [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+      [ Login-LAT-Port ]
+      [ Login-LAT-Service ]
+      [ Login-Service ]
+      [ Login-TCP-Port ]
+  *   [ Tunneling ]
+  *   [ Proxy-Info ]
+  *   [ Route-Record ]
+  *   [ AVP ]
+
+AC-Answer ::= <Diameter Header: 271, PXY>
```
Comment (RoadRunnr) [inline dia/diameter_nasreq_rfc7155.dia]: ACA?

File: dia/diameter_nasreq_rfc7155.dia (hunk 3 above)
Comment (surik) [inline dia/diameter_nasreq_rfc7155.dia]: I called it `AC-Request` like in RFC 7155 [3.9.  Accounting-Request (ACR) Command](https://tools.ietf.org/html/rfc7155#section-3.9). Should I rename to `ACR`?

File: .travis.yml (hunk 1 above)
Comment (surik) [inline .travis.yml]: See #2

File: dia/diameter_nasreq_rfc7155.dia (hunk 3 above)
Comment (RoadRunnr) [inline dia/diameter_nasreq_rfc7155.dia]: Yes, please. It seems to be common to use the short version give by the RFC (e.g. ACR)

File: dia/diameter_nasreq_rfc7155.dia (hunk 2 above)
Comment (surik) [inline dia/diameter_nasreq_rfc7155.dia]: done

File: src/ergw_aaa_diameter_client.erl [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+%% API
+%%===================================================================
+
+init(Opts) ->
+    OriginHost = proplists:get_value(origin_host, Opts, <<"client.example.com">>),
+    OriginRealm = proplists:get_value(origin_realm, Opts, <<"example.com">>),
+    SvcOpts = [{'Origin-Host', OriginHost},
+               {'Origin-Realm', OriginRealm},
+               {'Origin-State-Id', diameter:origin_state_id()},
+               {'Vendor-Id', 0},
```
Comment (RoadRunnr) [inline src/ergw_aaa_diameter_client.erl]: Vendor Id should be something sensible

Comment (RoadRunnr) [inline src/ergw_aaa_diameter_client.erl]: **IMPORTANT:** 'Host-IP-Address'** also needs to be included !!!

Comment (hwinkel) [inline src/ergw_aaa_diameter_client.erl]: could be a TP one or defines 3gpp one?

Comment (hwinkel) [inline src/ergw_aaa_diameter_client.erl]: Do we know which L4 protocol needs to be supported? I discusses with @vances that many providers and vendors don't even support SCTP for Diameter communication

File: rebar.config [hunk 6]
```diff
@@ -1,5 +1,7 @@
 {erl_opts, [debug_info, {parse_transform, lager_transform}]}.
 
+{plugins, [rebar3_diameter_compiler]}.
```
Comment (RoadRunnr) [inline rebar.config]: this is now contained twice in this file

Comment (surik) [inline rebar.config]: @RoadRunnr where is the second place?

Comment (RoadRunnr) [inline rebar.config]: line 13 (after the minimum_otp_vsn)

**Comment by RoadRunnr** [comment]
overall this looks nice, just a few minor tweaks to keep the style consistent across projects:

* remove the trailing white space (git diff shows them nicely in red)
* the indention and overall formatting of the DIAMETER dictionary should match that of the dictionaries shipped with Erlang
* many of the NASREQ DIAMETER attributes are reused from RADIUS, the writing of their enum values should match the way they are commonly used in RADIUS, e.g. 'Callback-NAS-Prompt' instead of CALLBACK_NAS_PROMPT, sync with the eradius dictionaries

**Pull request merged**